src/runtime_src/core/tools/xbmgmt2/OO_UpdateBase.cpp

```diff
@@ -0,0 +1,752 @@
+  if (is_versal) {
+    update_versal_SC(dev);
+    return;
+  }
+
+  //if factory image, update SC
+  auto is_mfg = xrt_core::device_query<xrt_core::query::is_mfg>(dev);
+  if (is_mfg) {
+    std::unique_ptr<firmwareImage> bmc = std::make_unique<firmwareImage>(file, BMC_FIRMWARE);
+    if (bmc->fail())
+      throw xrt_core::error(boost::str(boost::format("Failed to read %s") % file));
+
+    if (flasher.upgradeBMCFirmware(bmc.get()) != 0)
+      throw xrt_core::error("Failed to update SC flash image");
+    return;
+  }
+
+  // If SC is fixed, stop flashing immediately
+  if (is_SC_fixed(index))
+    throw xrt_core::error("SC is fixed, unable to flash image.");
+
+  // Mgmt pf needs to shutdown so that the board doesn't brick
+  try {
+    dev->device_shutdown();
+  }
+  catch (const xrt_core::error& e) {
+    throw xrt_core::error(std::string("Only proceed with SC update if all user applications for the target card(s) are stopped. ") + e.what());
+  }
+
+  std::unique_ptr<firmwareImage> bmc = std::make_unique<firmwareImage>(file, BMC_FIRMWARE);
+
+  if (bmc->fail())
+    throw xrt_core::error(boost::str(boost::format("Failed to read %s") % file));
+
+  if (flasher.upgradeBMCFirmware(bmc.get()) != 0)
+    throw xrt_core::error("Failed to update SC flash image");
+
+  // Bring back mgmt pf
+  try {
+    dev->device_online();
+  }
+  catch (const xrt_core::error& e) {
+    throw xrt_core::error(e.what() + std::string(" Please warm reboot."));
+  }
+
+  std::cout << boost::format("%-8s : %s \n\n") % "INFO" % "SC firmware image has been programmed successfully.";
+}
+
+// Helper function for header info
+static std::string
+file_size(const std::string & _file)
+{
+  std::ifstream in(_file.c_str(), std::ifstream::ate | std::ifstream::binary);
+  auto total_size = std::to_string(in.tellg());
+  int strSize = static_cast<int>(total_size.size());
+
+  //if file size is > 3 digits, insert commas after every 3 digits
+  static const int COMMA_SPACING = 3;
+  if (strSize > COMMA_SPACING) {
+   for (int i = (strSize - COMMA_SPACING);  i > 0;  i -= COMMA_SPACING)
+     total_size.insert(static_cast<size_t>(i), 1, ',');
+  }
+
+  return total_size + std::string(" bytes");
+}
+
+// Helper function for header info
+static std::pair<std::string, std::string>
+deployment_path_and_filename(std::string file)
+{
+  using tokenizer = boost::tokenizer< boost::char_separator<char> >;
+  boost::char_separator<char> sep("\\/");
+  tokenizer tokens(file, sep);
+  std::string dsafile = "";
+  for (auto tok_iter = tokens.begin(); tok_iter != tokens.end(); ++tok_iter) {
+  	if ((std::string(*tok_iter).find(XSABIN_FILE_SUFFIX) != std::string::npos)
+          || (std::string(*tok_iter).find(DSABIN_FILE_SUFFIX) != std::string::npos))
+          dsafile = *tok_iter;
+  }
+  auto pos = file.rfind('/') != std::string::npos ? file.rfind('/') : file.rfind('\\');
+  std::string path = file.erase(pos);
+
+  return std::make_pair(dsafile, path);
```

Comment:
```suggestion
  return { dsafile, path };
```
Difficult to understand what is done in this uncommented function, by the way...